src/main/scala/monocly/Optic.scala

```diff
@@ -1,65 +1,205 @@
 package monocly
 
 import monocly.impl._
-import monocly.functions.{Each, Index}
-import monocly.internal.{NonEmptyList, Applicative, Monoid}
+import monocly.functions.*
+import monocly.internal.*
+import OpticCan.*
 
 type Optic[+ThisCan, S, A] = POptic[ThisCan, S, S, A, A]
 
-final class POptic[+ThisCan, -S, +T, +A, -B] private[monocly](
-    protected[monocly] val getter: GetterImpl[ThisCan, S, A],
-    protected[monocly] val setter: SetterImpl[ThisCan, S, T, A, B]):
+final class POptic[+ThisCan, -S, +T, +A, -B] private[monocly](protected[monocly] val impl: OpticImpl[S, T, A, B]):
 
   def andThen[ThatCan >: ThisCan, C, D](o: POptic[ThatCan, A, B, C, D]): POptic[ThatCan, S, T, C, D] =
-    POptic(getter.andThen(o.getter), setter.andThen(o.setter))
+    POptic(impl.andThen(o.impl))
 
-end POptic
+  def each[C, S1 <: S, A1 >: A](using evEach: Each[A1, C])(using T <:< S1, A1 <:< B): Optic[ThisCan | (GetMany & Modify), S1, C] = 
```

Comment:
You right that the signature is very ugly and unnecessarily complex. Though I find it ok for `get`, `modify` etc when there is only one constraint based on `ThisCan`